src/bootman/bootman.c

```diff
@@ -446,6 +461,185 @@ bool boot_manager_set_uname(BootManager *self, const char *uname)
+                autofree(char) *initrd_source = NULL;
+
+                initrd_target = string_printf("%s%s/freestanding-%s",
+                                              base_path,
+                                              (is_uefi ? efi_boot_dir : ""),
+                                              (char*)self->initrd_freestanding->data[i]);
+                initrd_source = string_printf("%s/%s",
+                                              self->initrd_freestanding_dir,
+                                              (char*)self->initrd_freestanding->data[i]);
+                if (!cbm_files_match(initrd_source, initrd_target)) {
+                        if (!copy_file_atomic(initrd_source, initrd_target, 00644)) {
+                                LOG_FATAL("Failed to install initrd %s: %s",
+                                          initrd_target,
+                                          strerror(errno));
+                                return false;
+                        }
+                }
+        }
+        return true;
+}
+
+static bool boot_manager_in_initrd_freestanding(const BootManager *manager, const char *filename)
+{
+        for (int i = 0; i < manager->initrd_freestanding->len; i++) {
+                if (strcmp(filename, manager->initrd_freestanding->data[i]) == 0) {
+                        return true;
+                }
+        }
+        return false;
+}
+
+bool boot_manager_remove_initrd_freestanding(BootManager * self)
+{
+        autofree(char) *base_path = NULL;
+        autofree(char) *initrd_efi_path = NULL;
+        autofree(DIR) *initrd_dir = NULL;
+        struct dirent *ent = NULL;
+        bool is_uefi = ((self->bootloader->get_capabilities(self) & BOOTLOADER_CAP_UEFI) ==
+                        BOOTLOADER_CAP_UEFI);
+        const char *efi_boot_dir =
+            is_uefi ? self->bootloader->get_kernel_destination(self) : NULL;
+        if (!self || !self->initrd_freestanding_dir) {
+                return false;
+        }
+        /* if it's UEFI, then bootloader->get_kernel_dst() must return a value. */
+        if (is_uefi && !efi_boot_dir) {
+                return false;
+        }
+
+        base_path = boot_manager_get_boot_dir((BootManager *)self);
+
+        initrd_efi_path = string_printf("%s/%s",
+                                        base_path,
+                                        (is_uefi ? efi_boot_dir : ""));
+
+        initrd_dir = opendir(initrd_efi_path);
+        if (!initrd_dir) {
+                LOG_ERROR("Error opening %s: %s", initrd_efi_path, strerror(errno));
+                return false;
+        }
+
+        while ((ent = readdir(initrd_dir)) != NULL) {
+                autofree(char) *initrd_target = NULL;
+
+                if (strstr(ent->d_name, "freestanding-") != ent->d_name) {
+                        continue;
+                }
+
+                if (!boot_manager_in_initrd_freestanding(self, strchr(ent->d_name, '-') + 1)) {
```

Comment:
Since we are now searching over a list, this introduces an n^2 search. Could we move the initrd_freestanding to a hashmap instead?

Reply:
agree and changing